philander/stc311.py

```diff
@@ -0,0 +1,665 @@
+        # see alarmCallback from original implementation and max77960
+        pass
+
+    # Local functions for internal use
+
+    @staticmethod
+    def _transferSOC(data):
+        # LSB is 1/512 %, so shift by 9 bits.
+        ret = data + 0x0100 >> 9
+        return Percentage(ret)
+
+    @staticmethod
+    def _transferChangeRate(data):  # STC3117 exclusive
+        # LSB is 8.789 mC, scaling factor is 8.789 = 8789/1000
+        ret = (data * 8789 + 500) / 1000
+        return ret
+
+    @staticmethod
+    def _transferVoltage(data):
+        # LSB is 2.2mV, so scale by factor 2.20 = 22/10
+        ret = (data * 22 + 5) / 10
+        return Voltage(ret)
+
+    def _transferCurrent(self, data):
+        # Actually, we read out the voltage drop over the sense resistor.
+        # LSB is 5.88V, so first scaling factor is 5.88 = 294/50
+        # Value is signed!
+        # R = U/I  so we get I = U/R; Note that R is given in milliOhm!
+        # So, finally we scale by 294 / 50 * 1000 / rs = 294 * 20 / rs = 5880 / rs.
+        rs = self.REGISTER.CONFIG_GASGAUGE_0_RSENSE
+        if data >= 0:
+            ret = (data * 5880 + (rs / 2)) / rs
+        else:
+            ret = (data * 5880 - (rs / 2)) / rs
+        return ret
+
+    # /**
+    #  * AVG_CURRENT transfer function
+    #  */
+    def _transferCurrentAvg(self, data):  # STC3117 exclusive
+        if self.REGISTER.CHIP_TYPE is not ChipType.STC3117:
+            ret = Current.invalid  # Function is not implemented for this chip
+        else:
+            # Again, we actually read out the voltage drop over the sense resistor.
+            # LSB is 1.47V, partial scaling factor is 1.47 = 147/100
+            # Value is signed!
+            # Total scaling factor is: 147 / 100 * 1000 / rs = 147 * 10 / rs = 1470 / rs.
+            rs = self.REGISTER.CONFIG_GASGAUGE_0_RSENSE
+            if data >= 0:
+                ret = (data * 1470 + rs/2) / rs
+            else:
+                ret = (data * 1470 - rs/2) / rs
+            ret = Current(ret)
+        return ret
+
+    @staticmethod
+    def _transferTemperature(data):
+        # LSB is 1C, so we don't need any scaling.
+        return Temperature(data)
+
+    @staticmethod
+    def _transfertOCV(data):
+        # LSB is 0.55 mV, so scale by factor 0.55 = 55/100 = 11/20.
+        ret = (data * 11 + 10) / 20
+        return Voltage(ret)
+
+    @staticmethod
+    def _invTransferOCV(value):
+        # LSB is 0.55 mV, so scale by factor 1/0.55 = 100/55 = 20/11.
+        ret = (value * 20 + 5) / 11
+        return ret
+
+    @staticmethod
+    def _crc(data, length):
+        ret = 0
+        for idx in range(length):
+            ret ^= data[idx]
+        return ret
+
+    @staticmethod
+    def _checkRamConsistency(self, data, length):
+        # TODO: why is length never used? Refer to original implementation to see, what it does
+        # check if RAM test register value is correct
```

Comment:
Well observed! In Python, the array length is implicitly given with the array object.
The indices into the data[] array should be checked against the array length. Or, other way around, the array should be checked to have the minimum length needed.